### PR TITLE
Fix data corruption due to race condition

### LINKED ISSUE #86: Data corruption due to race condition
## Description
In jupyter-archive, reading and writing of `_write_buffer` is performed in different threads. So, for example, if a write to `_write_buffer` is performed between L1066 and L1067 in `RequestHandler.flush()`, buffer data will be lost.

Ref: https://github.com/tornadoweb/tornado/blob/master/tornado/web.py#L1066-L1067

I found that using `threading.Lock()` to lock `write()` and `flush()` solved the problem.
I will send PR if necessary.

## Reproduce

It is very difficult to reproduce this intentionally. If you archive several GB files, it will happen more often, but not all the time. 
In my application, I was able to reproduce it with high frequency for certain data, but I think it depends on the environment of application.

## Expected behavior
Data can be extracted without corruption.

## Context

- Python package version:

- Extension version:

- Operating System and its version: 
- Browser and its version: 

<details><summary>Command Line Output</summary>
<pre>
Paste the output from your command line running `jupyter lab` here, use `--debug` if possible.
</pre>
</details>

<details><summary>Web Browser Output</summary>
<pre>
Paste the output from your browser web console here.
</pre>
</details>

### PR DESCRIPTION
**Pull request by aidy1991**
Fix #86 

This PR solve a problem that archived data is corrupted because of race condition between buffer writing and buffer flushing.
I have found a way to reproduce this problem. It is a bit hacky, but it is possible to intentionally cause a race condition by modifying some of the handlers in tornado.

This problem is caused by other data being written to the buffer while the following two lines are executed.
So, inserting sleep between these two lines increases the probability that a race condition will occur.
https://github.com/tornadoweb/tornado/blob/master/tornado/web.py#L1066-L1067

Example:
```
        chunk = b"".join(self._write_buffer)
        time.sleep(0.1)
        self._write_buffer = []
```

In this PR, the `DownloadArchiveHandler` has a lock object, and uses that lock object when executing buffer wriging and buffer flushing.

### CODE REVIEW

File: jupyter_archive/handlers.py [hunk 1]
```diff
@@ -154,24 +158,28 @@ async def get(self, archive_path, include_body=False):
...
+            if self.canceled:
+                self.log.info("Download canceled.")
+            else:
+                # Here, we need to flush forcibly to move all data from _write_buffer to stream._write_buffer
+                self.flush(force=True)
+                self.log.info("Finished downloading {}.".format(archive_filename))
+        except Exception:
+            raise
+        finally:
+            self.flush_cb.stop()
```
Comment (aidy1991) [inline jupyter_archive/handlers.py:180]: In the investigation this problem, I noticed that callback thread is not stopped when an exception occurs during archive processing. This is a fix to ensure that callback thread is stopped.
This is not directly related to this issue, so please let me know if you prefer not to include it in this PR.

**Review by fcollonval — APPROVED** [review]
Thanks @aidy1991
